dev/detect and remove outliers.qmd

```diff
@@ -0,0 +1,276 @@
+---
+
+# Function `remove_outliers()`: detect and remove outliers based on speed
+
+This is a function developed to identify and remove outliers from GPS data using
+speed as the detecting metric. Speed is a good metric for outlier detection as
+it expresses distance covered given elapsed time, therefore highlighting
+instances of erroneous movement caused by sensor malfunctions. Here we have a
+close look at the function developed to perform this task.
+
+```{r}
+#| include: false
+#| warning: false
+#| message: false
+
+library(dplyr)
+library(ggplot2)
+library(move2)
+library(sf)
+library(units)
+
+source("../RFunction.R")
+source(file.path("../src/common/logger.R"))
+
+input1 <- readRDS("../data/raw/input1.rds")
+input2 <- readRDS("../data/raw/input2.rds")
+
+```
+
+
+Let's peak at the function's code.
+```{r}
+remove_outliers
+```
+
+
+Generate mock base data.
+```{r}
+n <- 10
+dt <- data.frame(
+  x = seq(30.01, 31.02, length.out = n), 
+  y = seq(2.01, 3.02, length.out = n),
+  time = seq.POSIXt(
+    as.POSIXct("2023-01-01 00:00:00 UTC"),
+    as.POSIXct("2023-01-01 01:30:00 UTC"), 
+    length.out = n
+  ), 
+  track = "a"
+)
+
+dt_mv2_base <- mt_as_move2(
+  dt,
+  coords = c("x", "y"), 
+  time_column = "time",
+  track_id_column = "track") |> 
+  sf::st_set_crs(4326L)
+
+dt_mv2_base
+```
+
+
+## Application to two consecutive outliers
+
+Add two outliers to the 2nd and 3rd positions and ad speeds (km/h).
+
+```{r}
+dt_mv2 <- dt_mv2_base
+st_geometry(dt_mv2)[c(2, 3)] <- st_sfc(st_point(c(12.11, 13.12)), st_point(c(12.4, 13.45)))
+dt_mv2$kmph <- mt_speed(dt_mv2) |> set_units("km/h") |> as.vector()
+dt_mv2
+```
+
+Plot locations and movement
+```{r}
+#| warning: false
+#| fig-cap-location: top
+#| fig-cap: Points are location events, dashed arrows illustrate movement paths and labels denote movement speeds (km/h)
+#| label: fig-p1
+
+ggplot(dt_mv2) +
+  geom_sf(col = "red", size = 3) +
+  geom_sf_label(aes(label = round(kmph, 1)), nudge_x = 0.8, size = 2.5) +
+  geom_curve(
+    aes(x = X, y = Y, xend = lead(X), yend = lead(Y)), 
+    data = as_tibble(st_coordinates(dt_mv2)),
+    curvature = -0.3,
+    arrow = arrow(length = unit(0.02, "npc"), type = "closed"),
+    linetype = "dashed"
+  )
+```
+
+Remove outliers.
+```{r}
+dt_mv2_clean <- dt_mv2 |> 
+  remove_outliers(kmph_tresh = 150)
+
+dt_mv2_clean <- dt_mv2_clean |> 
+  mutate(kmph = mt_speed(dt_mv2_clean) |> set_units("km/h") |> as.vector())
+
+dt_mv2_clean
+```
+
+Plot resultant data
+
+```{r}
+#| warning: false
+#| fig-cap-location: top
+#| fig-cap: Points are location events, dashed arrows illustrate movement paths and labels denote movement speeds (km/h)
+#| label: fig-p2
+
+ggplot(dt_mv2_clean) +
+  geom_sf(col = "red", size = 3) +
+  geom_sf_label(aes(label = round(kmph, 1)), nudge_x = 0.04, nudge_y = -0.04, size = 2.5) +
+  geom_curve(
+    aes(x = X, y = Y, xend = lead(X), yend = lead(Y)), 
+    data = as_tibble(st_coordinates(dt_mv2_clean)),
+    curvature = -0.3,
```

Comment:
not a comment but the curves lines are pretty :)

Reply:
Hi Bruno and Lindesay, thans for all the great coding here :) One idea that you have either already have implemented or not: We have previously adressed outlier removal by speed by doing a while loop until there are no more offending speeds, that way also e.g. groups of outliers are taken care of. THanks! Andrea

Reply:
Hi Andrea, thanks for having a look at this and for the feedback!

Yes, I think that's same/similar approach we take in our speed-outlier removal function :)

https://github.com/dmpstats/Standardise_Formats_and_Calculate_Basic_Statistics/blob/371ad170ca546aa4b7252085a3814994bb328a19/RFunction.R#L406-L463